src/drivers/gnss/septentrio/septentrio.cpp

```diff
@@ -0,0 +1,1439 @@
+
+#define SBF_CONFIG_RESET "setSBFOutput, Stream1, %s, none, off\n"
+
+#define SBF_CONFIG_RECEIVER_DYNAMICS "setReceiverDynamics, %s, UAV\n"
+
+#define SBF_CONFIG_ATTITUDE_OFFSET "setAttitudeOffset, %.3f, %.3f\n"
+
+#define SBF_TX_CFG_PRT_BAUDRATE 115200
+
+#define SBF_DATA_IO "setDataInOut, %s, Auto, SBF\n"
+
+#define SBF_CONFIG_RTCM_STATIC1 "" \
+	"setReceiverDynamics, Low, Static\n"
+
+#define SBF_CONFIG_RTCM_STATIC2 "" \
+	"setPVTMode, Static, , Geodetic1\n"
+
+#define SBF_CONFIG_RTCM_STATIC_COORDINATES "" \
+	"setStaticPosGeodetic, Geodetic1, %f, %f, %f\n"
+
+#define SBF_CONFIG_RTCM_STATIC_OFFSET "" \
+	"setAntennaOffset, Main, %f, %f, %f\n"
+
+static constexpr int SSN_SET_CLOCK_DRIFT_TIME_S{5}; ///< RTC drift time when time synchronization is needed (in seconds)
+
+SeptentrioGPS::SeptentrioGPS(const char *device_path) :
+	Device(MODULE_NAME)
+{
+	strncpy(_port, device_path, sizeof(_port) - 1);
+	// Enforce null termination.
+	_port[sizeof(_port) - 1] = '\0';
+
+	_report_gps_pos.heading = NAN;
+	_report_gps_pos.heading_offset = NAN;
+
+	int32_t enable_sat_info = 0;
+	param_get(param_find("SSN_SAT_INFO"), &enable_sat_info);
+
+	// Create satellite info data object if requested
+	if (enable_sat_info) {
+		_sat_info = new GPSSatelliteInfo();
+		_p_report_sat_info = &_sat_info->_data;
+		memset(_p_report_sat_info, 0, sizeof(*_p_report_sat_info));
+	}
+}
+
+SeptentrioGPS::~SeptentrioGPS()
+{
+	delete _dump_from_device;
+	delete _dump_to_device;
+	delete _rtcm_parsing;
+	delete _sat_info;
+}
+
+int SeptentrioGPS::print_status()
+{
+	PX4_INFO("status: %s, port: %s", _healthy ? "OK" : "NOT OK", _port);
+	PX4_INFO("sat info: %s", (_p_report_sat_info != nullptr) ? "enabled" : "disabled");
+	PX4_INFO("rate reading: \t\t%6i B/s", _rate_reading);
+
+	if (_report_gps_pos.timestamp != 0) {
+		PX4_INFO("rate position: \t\t%6.2f Hz", (double)get_position_update_rate());
+		PX4_INFO("rate velocity: \t\t%6.2f Hz", (double)get_velocity_update_rate());
+
+		PX4_INFO("rate publication:\t\t%6.2f Hz", (double)_rate);
+		PX4_INFO("rate RTCM injection:\t%6.2f Hz", (double)_rate_rtcm_injection);
+
+		print_message(ORB_ID(sensor_gps), _report_gps_pos);
+	}
+
+	return 0;
+}
+
+void SeptentrioGPS::run()
+{
+	uint64_t last_rate_measurement = hrt_absolute_time();
+	unsigned last_rate_count = 0;
+	param_t handle = param_find("SSN_YAW_OFFS");
+	float heading_offset = 0.f;
+
+	if (handle != PARAM_INVALID) {
+		param_get(handle, &heading_offset);
+		heading_offset = matrix::wrap_pi(math::radians(heading_offset));
+	}
+
+	if (initialize_communication_dump() == PX4_ERROR) {
+		SBF_WARN("GPS communication logging could not be initialized");
+	}
+
+	// Set up the communication, configure the receiver and start processing data until we have to exit.
+	while (!should_exit()) {
+		if (serial_open() != 0) {
+			continue;
+		}
+
+		decode_init();
+
+		// TODO: Not sure whether this is still correct when drivers are separate modules.
+		set_device_type(DRV_GPS_DEVTYPE_SBF);
```

Comment:
It's not very clear to me what the `set_device_type()` method is for. Is it only for drivers with multiple backends or should it be called here as well? What does the value represent?